src/engraving/dom/dynamic.cpp

```diff
@@ -423,19 +423,33 @@ void Dynamic::manageBarlineCollisions()
 //---------------------------------------------------------
 
 void Dynamic::setDynamicType(const String& tag)
+{
+    const DynamicType dt = parseDynamicText(tag);
+
+    if (dt == DynamicType::OTHER) {
+        LOGD("setDynamicType: other <%s>", muPrintable(tag));
+    }
+
+    setDynamicType(dt);
+}
+
+DynamicType Dynamic::parseDynamicText(const String& tag)
 {
     std::string utf8Tag = tag.toStdString();
-    size_t n = DYN_LIST.size();
-    for (size_t i = 0; i < n; ++i) {
-        if (TConv::toXml(DynamicType(i)).ascii() == utf8Tag || DYN_LIST[i].text == utf8Tag) {
-            setDynamicType(DynamicType(i));
-            setXmlText(String::fromUtf8(DYN_LIST[i].text));
-            return;
+    std::regex dynamicRegex(R"((?:<sym>.*?</sym>)+|(?:\b)[fmnprsz]+(?:\b(?=[^>]|$)))");
+    for (std::sregex_iterator it(utf8Tag.begin(), utf8Tag.end(), dynamicRegex), end; it != end; ++it) {
+        std::smatch match = *it;
+        std::string matchStr = match.str();
+        size_t n = DYN_LIST.size();
+        for (size_t i = 0; i < n; ++i) {
+            if (TConv::toXml(DynamicType(i)).ascii() == matchStr || DYN_LIST[i].text == matchStr) {
+                utf8Tag.replace(match.position(0), match.length(0), DYN_LIST[i].text);
+                setXmlText(String::fromStdString(utf8Tag));
```

Comment:
```suggestion
```
In principle this function should have a single responsibility - parsing and returning dynamics - so I don't think we should also be mutating the object with `setXmlText` here.

It might be easier said than done, but moving this call up to `setDynamicText` would be more cohesive. Once that's done we can make `parseDynamicText` a const function, and we can perhaps even make it private after the suggested changes to `textedit.cpp`.